Add support for WCS when animating line plots

Cadair opened the pull request:
I recently implemented astropy/astropy#9266 which enables the use of WCSAxes for 1D plots. This makes it much easier to provide a consistent plotting interface between 2D and 1D WCS aware plots. As part of an ongoing effort in sunpy/ndcube#207 I am trying to unify all the NDCube plotting to using WCS.

This PR implements a new `ArrayAnimatorWCS` class which allows you to plot 1D or 2D plots using a WCS, it provides a much reduced API interface, always relying on the WCS object for everything it can (no `axis_ranges` etc).

To make this work, I needed to change the base `ArrayAnimator` class to support (and always use underneath) functions to get the display coordinates for the sliders, this should be a backwards compatible change to the other animators.

TODO:

- [x] Error nicely if Astropy 4.0 isn't installed and 1D plots wont work.
- [x] Implement some method of setting WCSAxes properties on the coordinate object (format unit, label etc).
- [x] Work out how to deal with user overrides of `ylim` (@DanRyanIrish is this something we really need?)
- [x] Tests?! :scream: 
- [x] ~it would be nice to be able to set the display unit for the slider labels, but only because `astropy.wcs.WCS` forgets it and defaults to SI base.~


An animation of this working for an IRIS SJI cube (bottom slider is time, top slider is latitude):

![line_animator_2](https://user-images.githubusercontent.com/1391051/66818521-5af7b880-ef35-11e9-81b2-fac77f98d50a.gif)

pep8speaks commented:
Hello @Cadair! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`sunpy/visualization/animator/base.py`](https://github.com/sunpy/sunpy/blob/e43246afc831b4d142d04e4d1f6e5791ac8f5548/sunpy/visualization/animator/base.py):

> [Line 556:101](https://github.com/sunpy/sunpy/blob/e43246afc831b4d142d04e4d1f6e5791ac8f5548/sunpy/visualization/animator/base.py#L556): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (104 > 100 characters)
> [Line 561:101](https://github.com/sunpy/sunpy/blob/e43246afc831b4d142d04e4d1f6e5791ac8f5548/sunpy/visualization/animator/base.py#L561): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (116 > 100 characters)
> [Line 566:101](https://github.com/sunpy/sunpy/blob/e43246afc831b4d142d04e4d1f6e5791ac8f5548/sunpy/visualization/animator/base.py#L566): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (116 > 100 characters)
> [Line 595:1](https://github.com/sunpy/sunpy/blob/e43246afc831b4d142d04e4d1f6e5791ac8f5548/sunpy/visualization/animator/base.py#L595): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file
> [Line 611:101](https://github.com/sunpy/sunpy/blob/e43246afc831b4d142d04e4d1f6e5791ac8f5548/sunpy/visualization/animator/base.py#L611): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (110 > 100 characters)
> [Line 629:5](https://github.com/sunpy/sunpy/blob/e43246afc831b4d142d04e4d1f6e5791ac8f5548/sunpy/visualization/animator/base.py#L629): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 651:9](https://github.com/sunpy/sunpy/blob/e43246afc831b4d142d04e4d1f6e5791ac8f5548/sunpy/visualization/animator/base.py#L651): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)

dstansby commented:
Can you ping me for again when you've finished making changes 😛

Cadair commented:
Sure! I had hoped the tests would pass, but you know! :roll_eyes:

ayshih commented on sunpy/visualization/animator/image.py:
```diff
@@ -173,7 +173,9 @@ def __init__(self, data, wcs, image_axes=[-1, -2], unit_x_axis=None, unit_y_axis
-        super().__init__(data, image_axes=image_axes, axis_ranges=axis_ranges, **kwargs)
+
+        # For some reason using `super()` here causes an error.
+        ImageAnimator.__init__(self, data, image_axes=image_axes, axis_ranges=axis_ranges, **kwargs)
```
I don't expect you to know the answer since you say "for some reason" and "wtaf", but what's the deal?

Cadair replied:
I have no flaming idea, the error being raised was:
```
  File "/home/circleci/project/.tox/build_docs/lib/python3.7/site-packages/sunpy/visualization/animator/image.py", line 176, in __init__
    super().__init__(data, image_axes=image_axes, axis_ranges=axis_ranges, **kwargs)
TypeError: super(type, obj): obj must be an instance or subtype of type
```

dstansby replied:
Is this because of a funny interaction with the `@deprecated` dectorator? If you remove the dectorator does `super()` work?

Cadair replied:
You are correct, good spot! I have updated the comment.

DanRyanIrish commented on sunpy/visualization/animator/base.py:
```diff
@@ -532,30 +536,36 @@ def _sanitize_axis_ranges(self, axis_ranges, data_shape):
+        def get_pixel_to_world_callable(array):
+            def pixel_to_world(pixel):
+                return array[pixel]
+            return pixel_to_world
```
What is this code block doing...?

Cadair replied:
This function now converts all the slider axes to functions, which are then called by the base animator when the slider is updated. So this is just moving the array slice into a function.

DanRyanIrish replied:
Is there a reason it's a function defined in a function defined in a function?

Cadair replied:
yes, because it uses `array` from the defining scope (which is called a closure)

DanRyanIrish replied:
Still don't understand this but as I see you've thought it through a second time I trust it's fine ;)

DanRyanIrish commented on sunpy/visualization/animator/wcs.py:
```diff
@@ -0,0 +1,241 @@
+    def __init__(self, data, wcs, slices, coord_params=None, ylim=None, ylabel=None, **kwargs):
+        if not isinstance(wcs, BaseLowLevelWCS):
+            raise ValueError("A WCS object should be provided that implements the astropy WCS API.")
+        if wcs.pixel_n_dim is not data.ndim:
```
Should this not be ```wcs.pixel_n_dim =! data.ndim``` to keep it general to any number of dimensions?  (Although I realize if the dimensions were greater than the Python-pre-assigned ints the data cube will likely by too big for most systems...)

Cadair replied:
yes, yes it should.

DanRyanIrish commented on sunpy/visualization/animator/wcs.py:
```diff
@@ -0,0 +1,241 @@
+        image_axes = [slices[::-1].index("x")]
+        if "y" in slices:
+            image_axes.append(slices[::-1].index("y"))
+            self.plot_dimensionality = 2
```
Does the way this is done mean that you can't customize which dimensions are on the x and y axis for image animations?

Cadair replied:
It does not. You just specify `('y', 'x')` rather than `('x', 'y')` to specify the order.

DanRyanIrish commented on sunpy/visualization/animator/wcs.py:
```diff
@@ -0,0 +1,241 @@
+        based on the WCS.
+        """
+
+        def partial_pixel_to_world(pixel_dimension, pixel_coord):
```
What's the advantage of defining a function within a function here?

Cadair replied:
There isn't much either way really, it's just where the code belonged to me!

There is a potential performance optimisation in using closures here (for the lines in this function which are not dependant upon the pixel coord), which would require the function to be defined here, so I think it's probably a sane place for it.

DanRyanIrish replied:
OK. So long as there's a reason to justify its ugliness ;)

Cadair replied:
On further reflection I have moved it to its own method.

DanRyanIrish commented on sunpy/visualization/animator/wcs.py:
```diff
@@ -0,0 +1,241 @@
+            if grid is not None:
+                if not isinstance(grid, dict):
+                    grid = {}
+                coord.grid(**grid)
```
So the changes/actions of ```coord``` doesn't have to be done to the version attached to ```self```?

Cadair replied:
No, coord is attached to `self.axes` and coord is just the element in the dict, so this will happily work by reference.

DanRyanIrish replied:
So ```coord``` is a shallow copy of something attached to ```self.axes``` not a new deepcopy?

Cadair replied:
it's not even a copy really, it's just a variable name pointing at the object in the dict.

DanRyanIrish commented on sunpy/visualization/animator/wcs.py:
```diff
@@ -0,0 +1,241 @@
+        """
+        if val != slider.cval:
+            self.axes.reset_wcs(wcs=self.wcs, slices=self.slices_wcsaxes)
+            im.set_array(self.data_transposed)
```
So this means the data is re-transposed every time the plot is updated?  Is this not a performance concern?

Cadair replied:
It does, I don't think transpose is a particularly performance sensitive operation in numpy, as it doesn't modify the data, it just creates a transposed view into the array.

Cadair commented:
If you run the full test suite with figures you get 100% test coverage on astropy 4.0 :rofl: :sob:

Cadair commented:
@dstansby I am done with this, please review if you get a chance.

dstansby commented:
Looks like figure tests are failing.

Cadair commented:
The hashes just needed updating. Passes now.

dstansby commented:
Okay, I think that all makes sense - will give the PR a proper review now. Yes I think having an astropy-dev figure test run would be good (there's already one for none-figure tests right?)

dstansby commented on sunpy/visualization/animator/tests/test_basefuncanimator.py:
```diff
@@ -165,8 +169,8 @@ def test_sanitize_axis_ranges(axis_ranges, exp_extent, exp_axis_ranges):
     assert exp_extent == out_extent
-    assert np.array_equal(exp_axis_ranges[0], out_axis_ranges[0])
     assert np.array_equal(exp_axis_ranges[1], out_axis_ranges[1])
+    assert callable(out_axis_ranges[0])
```
Even though this is a callable now, can we still do a comparision with expected ranges (as in the old line 168 which has now gone)?

dstansby commented on sunpy/visualization/animator/tests/test_basefuncanimator.py:
```diff
@@ -206,5 +211,8 @@ def test_imageanimator_figure():
-    plt.show()
+    with warnings.catch_warnings():
+        warnings.simplefilter("ignore")
+        wcs_anim = ImageAnimatorWCS(sequence_array, wcs=wcs, vmax=1000, image_axes=[0, 1])
```
What warnings are being thrown up, and is it possible to do a more specific warning filter instead of just ignoring everything?

Cadair replied:
it was to filter the `SunPyDeprecationWarning` I will change it :+1:

dstansby commented on sunpy/visualization/animator/tests/test_wcs.py:
```diff
@@ -0,0 +1,155 @@
+@pytest.fixture
+def wcs_4d():
+    header = dedent("""\
+    WCSAXES =                    4 / Number of coordinate axes
```
👍 I approve of extensive commenting here

dstansby commented on sunpy/visualization/animator/tests/test_wcs.py:
```diff
@@ -0,0 +1,155 @@
+
+def test_constructor_errors(wcs_4d):
+    # WCS is not BaseLowLevelWCS
+    with pytest.raises(ValueError):
```
Could you add the `match='string that is in error message'` bit to `pytest.raises` here, so it's easier to read exactly what is being caught.

Cadair commented on sunpy/visualization/animator/line.py:
```diff
@@ -148,8 +148,7 @@ def update_plot(self, val, line, slider):
         """
         val = int(val)
         ax_ind = self.slider_axes[slider.slider_ind]
-        ind = int(np.argmin(np.abs(self.axis_ranges[ax_ind] - val)))
```
yes. I am not sure how this snuck in originally, but it was actually wrong and broke things.

DanRyanIrish commented:
Shifting the play button to the left has worked a treat!  One other request though.  In cases where the slider is in pixel units, can it say "pix" in the unit position to be explicit?

Cadair commented on sunpy/visualization/animator/base.py:
```diff
@@ -532,30 +536,36 @@ def _sanitize_axis_ranges(self, axis_ranges, data_shape):
                 # If axis range not supplied, set pixel center values as integers starting at 0.
-                axis_ranges[i] = np.arange(data_shape[i])
-            else:
+                axis_ranges[i] = get_pixel_to_world_callable(np.arange(data_shape[i]))
```
It's a change that was needed for the WCS to work, we need to be able to call `pixel_to_world` to get the world coordinate for the slider. I don't really want to go through the whole code and change it, also it's only the slider axes which are callables (the others are not).

(Also this is not a user facing change).

Cadair commented on sunpy/visualization/animator/base.py:
```diff
@@ -604,3 +617,16 @@ def edges_to_centers_nd(axis_range, edges_axis):
+
+def format_quantity_as_string(quantity):
+    """
+    An Astropy 3.1 compatibility wrapper for Quantity.to_string.
```
I have already added it to the list in #3154, will add a comment.

Cadair commented:
@dstansby I think I have addressed all your comments. I added a couple more commits following an in-person review/testing exercise with @DanRyanIrish just in an attempt to make it have better formatting and defaults.

Cadair commented:
The figure test comparison (https://22838-2165383-gh.circle-artifacts.com/0/root/project/figure_test_images/fig_comparison.html) seems to indicate it's all fine apart from the colorbar one which is borked.

DanRyanIrish commented:
Is there anything more you need from me on this PR, @Cadair?

Cadair commented:
Another review once I have finished with all the CI issues would be good, but not right now.

DanRyanIrish commented:
Cool.

Cadair commented on sunpy/tests/figure_hashes_py36.json:
```diff
@@ -28,5 +28,9 @@
     "sunpy.timeseries.tests.test_timeseriesbase.test_rhessi_peek": "fcd235c9627ad38afd1ca90605bbfed05e5c6fb299c3afcd5dced78de06a061b",
     "sunpy.visualization.animator.tests.test_basefuncanimator.test_imageanimator_figure": "43febed509078c145da4e2b78f8b1a8d5c90ed89a8ce880e8eaa73e2785b6e29",
-    "sunpy.visualization.animator.tests.test_basefuncanimator.test_lineanimator_figure": "22addc09bcee836322349c6f086b1305b8c1dabf4176fb89afbd70da83dc79c8"
+    "sunpy.visualization.animator.tests.test_basefuncanimator.test_lineanimator_figure": "360673f30fa79c3218b4962bafb0aff85ff29e05949a1c5f25a0d78d0f200aba",
```
yeah, it looks fine in the figure comparison page.

Cadair commented:
Ok, @DanRyanIrish @dstansby I think this is done.

Cadair commented:
I can write one, but the best way to use this is with `NDCube`. We are in a bit of a weird situation where this code is mostly used by ndcube but is currently in core. There are plans afoot to move it into ndcube at somepoint, probably around the time ndcube 2 is ready.

Cadair commented:
I found the old `ImageAnimatorWCS` example and updated it for this new class.

nabobalis commented on examples/plotting/imageanimatorwcs.py:
```diff
@@ -1,99 +0,0 @@
-# We can make sure the ticks do not overlap.
-solar_x.set_ticklabel(exclude_overlapping=True)
-# Set this one to false, so we can see the difference.
-solar_y.set_ticklabel(exclude_overlapping=False)
```
Why we deleting this?

Cadair replied:
Because it doesn't work, and I haven't implemented support for tick labels in `ArrayAnimatorWCS` as the new defaults in `WCSAxes` should generally be good enough.

Cadair replied:
(specifically all these settings do not survive updating the animation)

nabobalis replied:
Just remove that bit?
The old example was deleted for a reason.

Cadair replied:
What? I renamed this file to refelect the new name of the class.

nabobalis replied:
Then why did you say you just updated the old example?

Cadair replied:
Well I updated it and then renamed it, so it's very similar, just in a different file name.

nabobalis replied:
Ok that's fine. I thought you reverted to the old one.

Cadair commented:
Updated animation of using this thing in line mode:

![line_animator_3](https://user-images.githubusercontent.com/1391051/68770583-54389000-061e-11ea-845e-9d27088f3cfc.gif)

DanRyanIrish commented on sunpy/visualization/animator/base.py:
```diff
@@ -292,15 +298,15 @@ def _add_widgets(self):
+            locator = self.divider.new_locator(nx=2, ny=y, nx1=nx1)
             self.sliders[-1].set_axes_locator(locator)
-            sframe = widgets.Slider(self.sliders[-1], f"{i:d}",
+            sframe = widgets.Slider(self.sliders[-1], "",
```
What's going on in these lines and the ~10 above?

Cadair replied:
It is computing the layout for the sliders. I am not going to attempt to explain it in detail in a GH comment, but basically the whole thing works by generating a  x by y grid and then placing axes on the grid which span a given number of rows and columns. So calls to `new_locator` specify the start index for the axes and how many rows and columns it spans. (`nx1`, `ny1`).

DanRyanIrish commented on sunpy/visualization/animator/base.py:
```diff
@@ -292,15 +298,15 @@ def _add_widgets(self):
         self.slider_buttons = []
         for i in range(self.num_sliders):
-            x = i * 2
+            y = i * 2
```
Is changing this to y to fix one of the issues we identified when you were here?  Or is this another issue?

Cadair replied:
It's just because it's a row index not a column index, so it was confusing to me re-reading the code that it was `x` and not `y`.

Cadair commented:
I have added one more commit to the end of this, which changes the default behaviour for line plots, so that the global min/max of the array don't need to be computed by default. As this is likely to cause memory and/or compute issues when using larger than memory memmaps or dask arrays.

nabobalis commented:
Test failures all seem to be servers not letting us download data from them.
